backend/src/statistics/statistics.service.ts

```diff
@@ -0,0 +1,158 @@
+      peerSummary,
+      heatmapData,
+      networkData,
+    };
+  }
+
+  private async getAttemptSummaryByUserId(userId: number) {
+    const data = await this.prisma.roomSession.findMany({
+      where: { users: { some: { id: userId } }, endTime: { not: null } },
+      include: {
+        question: {
+          select: { difficulty: true },
+        },
+      },
+    });
+    const map = {
+      [Difficulty.EASY]: 0,
+      [Difficulty.MEDIUM]: 0,
+      [Difficulty.HARD]: 0,
+    };
+    for (const session of data) {
+      const { difficulty } = session.question;
+      map[difficulty] = map[difficulty] + 1;
+    }
+    return map;
+  }
+
+  private async getDurationSummaryByUserId(userId: number) {
+    const data = await this.prisma.roomSession.findMany({
+      where: {
+        users: { some: { id: userId } },
+        endTime: { not: null },
+        startTime: {
+          gte: subDays(new Date(), 30),
+        },
+      },
+      include: {
+        question: {
+          select: { difficulty: true },
+        },
+      },
+    });
+    return data
+      .filter((session) => !!session.endTime)
+      .map((session) => {
+        return {
+          difficulty: session.question.difficulty as Difficulty,
+          timetaken: differenceInSeconds(
+            session.endTime || new Date(), // Should be defined by selection.
+            session.startTime,
+          ),
+          date: session.startTime,
+        };
+      });
+  }
+
+  private async getPeerSummaryByUserId(userId: number) {
+    const data = await this.prisma.roomSession.findMany({
+      where: { users: { some: { id: userId } }, endTime: { not: null } },
+      include: {
+        question: {
+          select: { title: true },
+        },
+        users: true,
+      },
+    });
+    return data.map((session) => {
+      return {
+        userName: session.users.find((user) => user.id !== userId)?.name || "",
+        questionTitle: session.question.title,
+        date: session.startTime,
+      };
+    });
+  }
+
+  private async getHeatmapDataByUserId(userId: number) {
+    const data = await this.prisma.roomSession.findMany({
+      where: {
+        users: { some: { id: userId } },
+        endTime: { not: null },
+        startTime: {
+          gte: subYears(new Date(), 1),
+        },
+      },
+    });
+    return data.map((session) => {
+      return { date: session.startTime };
+    });
+  }
+
+  private async getNetworkDataByUserId(userId: number) {
+    const data = await this.prisma.roomSession.findMany({
+      where: {
+        users: { some: { id: userId } },
+        endTime: { not: null },
+      },
+      include: {
+        question: {
+          include: {
+            topics: true,
+          },
+        },
+      },
+    });
+    const topics = new Map<number, string>();
+    const relations = new Map<number, number>();
+    for (const session of data) {
+      const sortedTopics = session.question.topics.sort((l, r) => l.id - r.id);
```

Comment:
Can we be more verbose with l and r to avoid magic strings? I suppose l stands for left and r stands for right

Reply:
On a side note, is there a reason why we need to sort the topics?

Reply:
It reduces the run time of the processing since we always point from the smaller ID to the larger ID.

Reply:
Noted